File: aribook/_template_aribook231DPナップサック問題.cpp
// aribook231 DP ナップサック問題
#include<iostream>
#include<algorithm>
//#include <numeric>
//#include<string>
//#include<vector>
//#include<map>
//#include<tuple>
//#include<queue>
//#include<regex>

// #include <bits/stdc++.h>
using namespace std;

const int MAX_N = 100;
const int MAX_W = 10000;
// 入力
int n, W;
int w[MAX_N], v[MAX_N];
int dp[MAX_N + 1][MAX_W + 1];	// メモ化テーブル

int main() {
	cin >> n >> W;
	for (int i = 0; i < n; i++)	{
		cin >> w[i] >> v[i];
	}

	// 動的計画法(DP)
	for (int i = n - 1; i >= 0; i--)
	{
		for (int j = 0; j <= W; j++)
		{
			if (j < w[i])
			{
				dp[i][j] = dp[i + 1][j];
			}
			else
			{
				dp[i][j] = max(dp[i + 1][j], dp[i + 1][j - w[i]] + v[i]);
			}
		}
	}
	cout << dp[0][W] << endl;

	return 0;
}

File: aribook/_template_aribook231DPナップサック問題.py
# aribook231 DP ナップサック問題

import numpy as np

MAX_N = 100
MAX_W = 10000

N, W = map(int, input().split())
w, v = [], []
for i in range(N):
    w_, v_ = map(int, input().split())
    w.append(w_)
    v.append(v_)

# i番目以降の品物から重さの総和がj以下となるように選ぶ
dp = np.zeros((N+1, W+1), np.int)

for i in range(N-1, -1, -1):
    for j in range(W+1):
        if j < w[i]:
            dp[i, j] = dp[i+1, j]
        else:
            dp[i, j] = max(dp[i+1, j], dp[i+1, j-w[i]] + v[i])

print(dp[0, W])


"""
input
N W
w1 v1
wn vn

testdata
4 5
2 3
1 2
3 4
2 2

output
7
"""
